modules/dnn/src/layers/scale_layer.cpp

```diff
@@ -412,15 +414,20 @@ class DataAugmentationLayerImpl CV_FINAL : public DataAugmentationLayer
         float* outData = outputs[0].ptr<float>();
 
         Mat data_mean_cpu = blobs[1].clone();
+        Mat mean_resize = Mat(inputs[0].size[3], inputs[0].size[2], CV_32FC3);
+        Mat mean_3d = Mat(data_mean_cpu.size[3], data_mean_cpu.size[2], CV_32FC3, data_mean_cpu.ptr<float>(0));
+        resize(mean_3d, mean_resize, Size(inputs[0].size[3], inputs[0].size[2]));
+        int new_size[] = {1, mean_resize.channels(), mean_resize.cols, mean_resize.rows};
+        Mat data_mean_cpu_resize = mean_resize.reshape(1, *new_size);
```

Comment:
Could you provide a link to this code from original implementation?

Reply:
Actucally I think the original implementation ([code from caffe](https://github.com/lmb-freiburg/flownet2/blob/9eed763789c06fa5f44e7af79078c0446a82bfc6/src/caffe/layers/data_augmentation_layer.cu#L604)) does not consider the case and is potentially a bug.
In the original implementation, `count` is calculated by `count = cropped_width_*cropped_height_*bottomchannels` which is basically the `input[0]`'s shape in our case. However, `data_mean_gpu` could have different sizes against `top_data`/`bottom_data`. In the original implementation, the authors skipped this and no error occurred because of `data_mean_gpu` is `float*` in the cuda code. However, in our case, we have to explicitly resize the `data_mean_cpu` to match the shape of the `input[0]`.

Reply:
To be sure of this approach, we need to add a test. Create a simple proto with the appropriate params and compare the result of Caffe and OpenCV.

Reply:
Sure. I will have a try these days when my other work is done. Also, could you suggest any example so that I can learn from the test code style since this is my first time to contribute to opencv?

Reply:
1. Create proto, run net in Caffe and save inputs and outputs using numpy. Open PR to `opencv_extra`. For example, (https://github.com/opencv/opencv_extra/pull/708)
2. Add test https://github.com/opencv/opencv/blob/3.4/modules/dnn/test/test_layers.cpp#L644

Reply:
Hi @l-bat , I am most finishing the testing. However, there is one thing that I may have to confirm with you first. I spotted a minor difference between the opencv's reimpl and the Caffe's code. To be specific, in https://github.com/lmb-freiburg/flownet2/blob/master/src/caffe/layers/data_augmentation_layer.cu#L353, the `num_iter` is 1 at the very beginning and is increased every iter once the network is built. However, in https://github.com/opencv/opencv/blob/master/modules/dnn/src/layers/scale_layer.cpp#L409, `num_iter` is set to be 0 for each image pair.
There are basically two differences: (1) for Caffe, the `num_iter` starts from 1 while for OpenCV, the `num_iter` starts from 0. Since the param `recompute_mean: 1` in the test proto. The original Caffe is simply skipping the recomputation stage while OpenCV is doing the recomputation (akin to the "self-normalization", which is basically calculating the stats of the input image itself.)
(2) Caffe supports `recompute_mean = N > 1`, which is basically moving-averaging the stats of the testing images when there are more than one pair. However, it seems that OpenCV cannot achieve this.
What do you think? I think the current approach of OpenCV makes sense as well (i.e. "self-normalization"). If we agree on this, I can modify the Caffe code to fit our style and obtain the result in order to pass the test.

Reply:
@l-bat A friendly reminder.

Reply:
Sorry for the late reply. In Opencv `num_iter` increases in https://github.com/opencv/opencv/blob/master/modules/dnn/src/layers/scale_layer.cpp#L420. So, in OpenCV `num_iter = 1`  at the very beginning just like in Caffe.

We conducted several experiments with the DataAugmentation layer. In our experiments in `Forward_gpu` `num_iter` will always reinitialized and set to 1. We think this is bug in Сaffe implementation. We try to fix this, but when we increment `num_iter` visual results has become worse. I think in Caffe and OpenCV `num_iter` is set to be 1 for each image pair.
If this is not right please provide the model where num_iter != 1 and we will rewrite `forward` in OpenCV.

Why do you think Caffe doesn't support `recompute_mean = 1`? I only found this check: https://github.com/lmb-freiburg/flownet2/blob/master/src/caffe/layers/data_augmentation_layer.cu#L593
In OpenCV we get `recompute_mean` from proto (default value = 1) and check `recompute_mean > 0`: https://github.com/opencv/opencv/blob/master/modules/dnn/src/layers/scale_layer.cpp#L382

Reply:
Thanks for the detailed explanation.
> In our experiments in Forward_gpu num_iter will always reinitialized and set to 1. We think this is bug in Сaffe implementation. We try to fix this, but when we increment num_iter visual results has become worse.

Yes, indeed. Based on https://github.com/opencv/opencv/blob/master/modules/dnn/src/layers/scale_layer.cpp#L409 and https://github.com/opencv/opencv/blob/master/modules/dnn/src/layers/scale_layer.cpp#L420, `num_iter=1`.
 However in Caffe, from https://github.com/lmb-freiburg/flownet2/blob/master/src/caffe/layers/data_augmentation_layer.cu#L353, I can see `num_iter=1`. Because of an additional line `num_iter = ((int)num_iter+1);` (in https://github.com/lmb-freiburg/flownet2/blob/master/src/caffe/layers/data_augmentation_layer.cu#L354). I think this means `num_iter` is actually 2, which is why I am saying that "Caffe doesn't support `recompute_mean = 1`." since Caffe is always skipping the recomputation (i.e. re-using the stats in the training phase). See: https://github.com/lmb-freiburg/flownet2/blob/master/src/caffe/layers/data_augmentation_layer.cu#L600
I think your current re-implementation is correct :)